sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -163,133 +190,114 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
-        q"""
-          val $nullTerm = ${value == null}
-          val $primitiveTerm: ${termForType(dataType)} = $value
-         """.children
-
-      case Cast(e @ BinaryType(), StringType) =>
-        val eval = expressionEvaluator(e)
-        eval.code ++
-        q"""
-          val $nullTerm = ${eval.nullTerm}
-          val $primitiveTerm =
-            if($nullTerm)
-              ${defaultPrimitive(StringType)}
-            else
-              org.apache.spark.sql.types.UTF8String(${eval.primitiveTerm}.asInstanceOf[Array[Byte]])
-        """.children
+        s"""
+          final boolean $nullTerm = true;
+          ${primitiveForType(dataType)} $primitiveTerm = ${defaultPrimitive(dataType)};
+        """
+
+      case expressions.Literal(value: UTF8String, StringType) =>
+        val arr = s"new byte[]{${value.getBytes.map(_.toString).mkString(", ")}}"
+        s"""
+          final boolean $nullTerm = false;
+          org.apache.spark.sql.types.UTF8String $primitiveTerm =
+            new org.apache.spark.sql.types.UTF8String().set(${arr});
+         """
+
+      case expressions.Literal(value, FloatType) =>
+        s"""
+          final boolean $nullTerm = false;
+          float $primitiveTerm = ${value}f;
+         """
+
+      case expressions.Literal(value, dataType) =>
+        s"""
+          final boolean $nullTerm = false;
+          ${primitiveForType(dataType)} $primitiveTerm = $value;
+         """
+
+      case Cast(child @ BinaryType(), StringType) =>
+        child.castOrNull(c =>
+          s"new org.apache.spark.sql.types.UTF8String().set($c)",
+          StringType)
 
       case Cast(child @ DateType(), StringType) =>
         child.castOrNull(c =>
-          q"""org.apache.spark.sql.types.UTF8String(
+          s"""new org.apache.spark.sql.types.UTF8String().set(
                 org.apache.spark.sql.catalyst.util.DateUtils.toString($c))""",
           StringType)
 
+      case Cast(child @ BooleanType(), dt: NumericType) =>
+        child.castOrNull(c => s"(${primitiveForType(dt)})($c?1:0)", dt)
+
+      case Cast(child @ NumericType(), ByteType) =>
+        child.castOrNull(c => s"(byte)($c)", ByteType)
+
+      case Cast(child @ NumericType(), ShortType) =>
+        child.castOrNull(c => s"(short)($c)", ShortType)
+
       case Cast(child @ NumericType(), IntegerType) =>
-        child.castOrNull(c => q"$c.toInt", IntegerType)
+        child.castOrNull(c => s"(int)($c)", IntegerType)
 
       case Cast(child @ NumericType(), LongType) =>
-        child.castOrNull(c => q"$c.toLong", LongType)
+        child.castOrNull(c => s"(long)($c)", LongType)
 
       case Cast(child @ NumericType(), DoubleType) =>
-        child.castOrNull(c => q"$c.toDouble", DoubleType)
+        child.castOrNull(c => s"(double)($c)", DoubleType)
 
       case Cast(child @ NumericType(), FloatType) =>
-        child.castOrNull(c => q"$c.toFloat", FloatType)
+        child.castOrNull(c => s"(float)($c)", FloatType)
```

Comment:
This's probably not true for `Decimal` (of child), we still need to call `$c.toFloat` etc.